README.md
# simple_cluster

Simple clustering library implemented in Dart.

## Getting Started
In your Dart (or Flutter) project `pubspec.yaml` add the dependency:
```yaml
dependencies:
  ...
  simple_cluster: ^[latest version]
```

## Usage
### DBSCAN
Density-based spatial clustering of applications with noise (DBSCAN) [Check Wiki](https://en.wikipedia.org/wiki/DBSCAN).

Parameter:
- `epsilon`: Maximum distance for two point to be considered as cluster
- `minPoints`: Minimum points in a single cluster
- `distanceMeasure`: Distance measure between 2 points

Example in dart.
```dart
import 'package:simple_cluster/src/dbscan.dart';

void main() {
  List<List<double>> dataset = [
    [1, 2], [2, 2], [2, 3], [8, 7], [8, 8], [25, 80]
  ];

  List<List<double>> dataset2 = [
    [1,1],[0,1],[1,0],
    [10,10],[10,13],[13,13],
    [54,54],[55,55],[89,89],[57,55]
  ];

  DBSCAN dbscan = DBSCAN(
    epsilon: 3,
    minPoints: 2,
  );

  List<List<int>> clusterOutput = dbscan.run(dataset);
  print("===== 1 =====");
  print("Clusters output");
  print(clusterOutput);//or dbscan.cluster
  print("Noise");
  print(dbscan.noise);
  print("Cluster label for points");
  print(dbscan.label);

  List<List<int>> clusterOutput2 = dbscan.run(dataset2);
  print("===== 2 =====");
  print("Clusters output");
  print(clusterOutput2);//or dbscan.cluster
  print("Noise");
  print(dbscan.noise);
  print("Cluster label for points");
  print(dbscan.label);
}

// Output
// ===== 1 =====
// Clusters output
// [[0, 1, 2], [3, 4]]
// Noise
// [5]
// Cluster label for points
// [0, 0, 0, 1, 1, -1]
// ===== 2 =====
// Clusters output
// [[0, 1, 2], [3, 4, 5], [6, 7, 9]]
// Noise
// [8]
// Cluster label for points
// [0, 0, 0, 1, 1, 1, 2, 2, -1, 2]
```

## Hierarchical
Hierarchical clustering [Check Wiki](https://en.wikipedia.org/wiki/Hierarchical_clustering).

Parameter:
- `maxLinkage`: Terminate clustering process based on linkage
- `minCluster`: Terminate clustering process based number of cluster
- `distanceMeasure`: Distance masure between 2 points
- `linkage`: Use existing linkage method
- `linkageMeasure`: Use custom linkage method

Example in dart.
```dart
  List<List<double>> colors = [
    [20, 20, 80],
    [22, 22, 90],
    [250, 255, 253],
    [100, 54, 255]
  ];

  List<List<double>> singles = [[7], [10], [20], [28], [35]];

  Hierarchical hierarchical = Hierarchical(
    minCluster: 2, //stop at 2 cluster
    linkage: LINKAGE.SINGLE
  );

  List<List<int>> clusterList = hierarchical.run(colors);

  print("===== 1 =====");
  print("Clusters output");
  print(clusterList);//or hierarchical.cluster
  print("Noise");
  print(hierarchical.noise);
  print("Cluster label for points");
  print(hierarchical.label);

  List<List<int>> clusterList2 = hierarchical.run(singles);

  print("===== 2 =====");
  print("Clusters output");
  print(clusterList2);//or hierarchical.cluster
  print("Noise");
  print(hierarchical.noise);
  print("Cluster label for points");
  print(hierarchical.label);

  // Output:
  // ===== 1 =====
  // Clusters output
  // [[2], [3, 1, 0]]
  // Noise
  // [2]
  // Cluster label for points
  // [0, 0, -1, 0]
  // ===== 2 =====
  // Clusters output
  // [[1, 0], [4, 3, 2]]
  // Noise
  // []
  // Cluster label for points
  // [0, 0, 1, 1, 1]
```
# Remarks
This library is inspired by:
- https://github.com/uhho/density-clustering
- https://github.com/math-utils/hierarchical-clustering

CHANGELOG.md
## [0.3.0]
- Null-safety support (thanks to stasgora)

## [0.2.1]
- Fixed hierarchical labeling

## [0.2.0+1]
- Added missing export

## [0.2.0]
- Added Hierarchical clustering
- Moved euclidDistance to common file

## [0.1.2]
- Updated License

## [0.1.1]
- Updated multiple location for pub.dev health

## [0.1.0] - Initial Release
- Added DBSCAN clustering


example/test.dart
// class UnableToConnectException implements Exception {
//   String message() => "Unable to connect. Please check your internet.";
// }

// void main(){
//   try {
//     run();
//   } catch (e){
//     print(e.message());
//   }
// }

// void run(){
//   throw Exception("waaa");
// }

// class CustomException implements Exception {
//   String _message;

//   String get message {
//     return _message;
//   }

//   CustomException([this._message = "Error message."]);
// }

// class UnableToConnectException extends CustomException {
//   String _message = "Unable to connect. Please check your internet.";
// }

// void main(){
//   try {
//     run();
//   } catch (e){
//     print(e.message);
//   }
// }

// void run(){
//   throw UnableToConnectException();//work, show Unable to connect. Please check your internet.
//   throw UnableToConnectException("wa");//not
//   throw CustomException();//work, show Error message.
//   throw CustomException("wa");//work, show wa
// }

void main(){
  String img = '<img src="https://github.com/flutter/website/raw/master/src/images/homepage/reflectly-hero-600px.png" alt="Reflectly hero image" style="max-width:100%;">';

  RegExp regExp = RegExp(r'<img.*src="(.*?)".*>');

  Match match = regExp.firstMatch(img);
  print(match.group(1));
}


lib/src/dbscan.dart
import 'package:simple_cluster/src/common.dart';

///Density-based spatial clustering of applications with noise (DBSCAN)
class DBSCAN {
  ///Complete list of points (in vector form)
  late List<List<double>> dataset;

  ///Result clusters
  List<List<int>> _cluster = [];

  ///Index of points considered as noise
  List<int> _noise = [];

  ///Cluster label, if prefer sklearn structure of output
  List<int>? _label;

  ///Threshold distance for two points to be considered as neighbor
  final double epsilon;

  ///Minimum points in neighborhood to be considered as a cluster
  final int minPoints;

  ///Distance measurement between two points
  double Function(List<double>, List<double>) distanceMeasure;

  ///Cluster label
  int? _currentLabel;

  DBSCAN({
    this.epsilon = 1,
    this.minPoints = 2,
    this.distanceMeasure = euclideanDistance,
  });
  
  ///Index of points considered as noise
  List<int> get noise {
    return _noise;
  }

  ///Result clusters
  List<List<int>> get cluster {
    return _cluster;
  }

  ///Cluster label for each points, if prefers sklearn's output structure.
  ///
  ///-1 means noise (doesn't belong in any cluster)
  List<int>? get label {
    return _label;
  }

  ///Run clustering process, add configs in constructor
  List<List<int>> run(List<List<double>> dataset) {
    if (dataset == null) {
      throw new Exception("Dataset must not be null");
    }

    //save dataset to class' variable
    this.dataset = dataset;

    //initial variables
    _cluster = [];
    _currentLabel = -1;
    _label = List.generate(dataset.length, (_) => -1);

    for (int i = 0; i < dataset.length; i++) {
      //skip labeled points
      if (_label![i] != -1) continue;

      //neighbor indexes
      List<int> neighbors = _rangeQuery(dataset[i]);

      if (neighbors.length < minPoints) {
        _noise.add(i);
      } else {
        //add new cluster
        List<int> newCluster = [];
        _currentLabel = _currentLabel! + 1;
        _cluster.add(newCluster);
        _expandCluster(i, neighbors, newCluster);
      }
    }

    return _cluster;
  }

  ///Expand cluster by checking neighbor's neighbor.
  ///
  ///Mutates parameter `neighbors` and `cluster`.
  void _expandCluster(int pointIndex, List<int> neighbors, List<int> cluster) {
    _addToCluster(pointIndex, _currentLabel!);

    for (int i = 0; i < neighbors.length; i++) {
      //Change Noise to border point
      if (_noise.contains(neighbors[i])) {
        _noise.remove(neighbors[i]);
        _addToCluster(neighbors[i], _currentLabel!);
      }

      //skip labeled points
      if (_label![neighbors[i]] != -1) continue;

      //add point to cluster
      _addToCluster(neighbors[i], _currentLabel!);

      //expand neighborhood
      List<int> expandedNeighbors = _rangeQuery(dataset[neighbors[i]]);

      if (expandedNeighbors.length >= minPoints) {
        neighbors = _joinIndexList(neighbors, expandedNeighbors);
      }
    }
  }

  ///Add to cluster and set label to point
  void _addToCluster(int pointIndex, int clusterLabel) {
    _cluster[clusterLabel].add(pointIndex);
    _label![pointIndex] = clusterLabel;
  }

  ///Return all point's index within p's eps-neighborhood (including p)
  List<int> _rangeQuery(List<double> p) {
    List<int> neighbors = [];
    for (int i = 0; i < dataset.length; i++) {
      if (distanceMeasure(p, dataset[i]) <= epsilon) {
        neighbors.add(i);
      }
    }

    return neighbors;
  }

  List<int> _joinIndexList(List<int> list1, List<int> list2) {
    List<int> newList = [...list1];

    for (int i = 0; i < list2.length; i++) {
      if (!newList.contains(list2[i])) {
        newList.add(list2[i]);
      }
    }

    return newList;
  }
}


lib/src/hierarchical.dart
import 'dart:math';

import 'package:simple_cluster/src/common.dart';

enum LINKAGE { SINGLE, COMPLETE, AVERAGE }

class Level {
  ///linkage between clusters
  double linkage;

  ///clusters in current level
  List<List<int>> cluster;

  ///In this level, cluster index 'joinFrom' are joined into 'joinTo'
  int? joinFrom;

  ///In this level, cluster index 'joinFrom' are joined into 'joinTo'
  int? joinTo;

  Level({this.linkage = 0, required this.cluster, this.joinFrom, this.joinTo});
}

class MinLinkage {
  ///Linkage value
  double linkage;

  ///MinLinkage is between cluster index i & j
  int i;

  ///MinLinkage is between cluster index i & j
  int j;

  MinLinkage({required this.i, required this.j, required this.linkage});
}

class Hierarchical {
  ///Complete list of points (in vector form)
  late List<List<double>> dataset;

  ///Resulting level
  late List<Level> _level;

  ///Index of points considered as noise
  List<int> _noise = [];

  ///Cluster label, if prefer sklearn structure of output
  List<int> _label = [];

  ///Terminate clustering process based on linkage
  final double maxLinkage;

  ///Terminate clustering process based number of cluster
  final int minCluster;

  ///Distance measurement between two points.
  ///Default is Euclidean
  double Function(List<double>, List<double>) distanceMeasure;

  ///Linkage type
  LINKAGE linkage;

  ///Linkage measurement
  double Function(List<double> distanceList)? linkageMeasure;

  ///Distance matrix between input
  late List<List<double>> _distanceMatrix;

  ///Current level's cluster
  List<List<int>> _cluster = [];
  Map<String, double> _cache = {};

  ///Resulting level
  List<Level> get level {
    return _level;
  }

  ///Index of points considered as noise
  List<int> get noise {
    return _noise;
  }

  ///Cluster label, if prefer sklearn structure of output
  List<int> get label {
    return _label;
  }

  ///Cluster of final level
  List<List<int>> get cluster {
    return _level[_level.length - 1].cluster;
  }

  Hierarchical({
    this.maxLinkage = double.infinity,
    this.minCluster = 1,
    this.distanceMeasure = euclideanDistance,
    this.linkage = LINKAGE.COMPLETE,
    this.linkageMeasure,
  }) {
    //if linkageMeasure is not overriden
    if (linkageMeasure == null) {
      switch (linkage) {
        case LINKAGE.AVERAGE:
          this.linkageMeasure = averageLink;
          break;
        case LINKAGE.COMPLETE:
          this.linkageMeasure = completeLink;
          break;
        case LINKAGE.SINGLE:
          this.linkageMeasure = singleLink;
          break;
      }
    }
  }

  ///Run clustering process and returns list of cluster
  List<List<int>> run(List<List<double>> dataset) {
    this.dataset = dataset;

    ///Initial configuration before running process
    buildDistanceMatrix();
    _cache = {};

    _cluster = [];
    //for every dataset, add it into their own cluster
    for (int i = 0; i < dataset.length; i++) {
      _cluster.add([i]);
    }

    //add first level
    Level currentLevel = Level(cluster: _cluster);
    _level = [currentLevel];

    while (_cluster.length > minCluster && currentLevel.linkage < maxLinkage) {
      currentLevel = mergeCluster();
    }

    //add other info
    _label = List.generate(dataset.length, (_) => -1);
    _noise = [];
    int currentLabel = 0;
    for (int i = 0; i < _cluster.length; i++) {
      if (_cluster[i].length > 1) {
        for (int j = 0; j < _cluster[i].length; j++) {
          _label[_cluster[i][j]] = currentLabel;
        }
        currentLabel++;
      } else {
        //if a cluster contain one member, consider it as noise
        _noise.add(_cluster[i][0]);
      }
    }

    return _cluster;
  }

  ///Merging two clusters.
  ///
  ///This is a process done for every hierarchical clustering level.
  Level mergeCluster() {
    MinLinkage? min;
    for (int i = 0; i < _cluster.length; i++) {
      for (int j = 0; j < i; j++) {
        double linkage = getLinkage(_cluster[i], _cluster[j]);

        if (min == null || linkage < min.linkage) {
          min = MinLinkage(i: i, j: j, linkage: linkage);
        }
      }
    }

    //create new cluster from old cluster (deep copy)
    List<List<int>> cluster = _cluster
        .map((oneClus) => oneClus.map((index) => index).toList())
        .toList();
    //join cluster index j into i
    cluster[min!.i].addAll(cluster[min.j]);
    //and remove cluster index j
    cluster.removeAt(min.j);
    Level currentLevel = Level(
        linkage: min.linkage, cluster: cluster, joinFrom: min.j, joinTo: min.i);

    _level.add(currentLevel);
    _cluster = cluster;

    return currentLevel;
  }

  ///Calculate linkage between 2 clusters
  ///
  ///Clusters contains index of dataset
  double getLinkage(List<int> cluster1, List<int> cluster2) {
    String hash = cluster1.length > cluster2.length
        ? "$cluster1-$cluster2"
        : "$cluster2-$cluster1";

    if (_cache.containsKey(hash)) return _cache[hash]!;

    List<double> distanceList = [];

    for (int i = 0; i < cluster1.length; i++) {
      for (int j = 0; j < cluster2.length; j++) {
        //get dataset index from the cluster
        distanceList.add(getDistance(cluster1[i], cluster2[j]));
      }
    }

    return _cache[hash] = linkageMeasure!(distanceList);
  }

  ///Get distance between 2 point in dataset
  ///
  ///i & j is the dataset's index
  double getDistance(int i, int j) {
    if (i > j) return _distanceMatrix[i][j];
    return _distanceMatrix[j][i];
  }

  ///Build distance matrix (only 1/2 filled)
  void buildDistanceMatrix() {
    List<List<double>> matrix2d = [];

    for (int i = 0; i < dataset.length; i++) {
      List<double> vector = [];
      for (int j = 0; j < i; j++) {
        vector.add(distanceMeasure(dataset[i], dataset[j]));
      }
      matrix2d.add(vector);
    }

    _distanceMatrix = matrix2d;
  }

  double singleLink(List<double> distanceList) {
    return distanceList.reduce(min);
  }

  double completeLink(List<double> distanceList) {
    return distanceList.reduce(max);
  }

  double averageLink(List<double> distanceList) {
    double sum = distanceList.reduce((a, b) => a + b);
    return sum / distanceList.length;
  }
}


example/dbscan_sample.dart
import 'package:simple_cluster/src/dbscan.dart';

void main() {
  List<List<double>> dataset = [
    [1, 2], [2, 2], [2, 3], [8, 7], [8, 8], [25, 80]
  ];

  List<List<double>> dataset2 = [
    [1,1],[0,1],[1,0],
    [10,10],[10,13],[13,13],
    [54,54],[55,55],[89,89],[57,55]
  ];

  DBSCAN dbscan = DBSCAN(
    epsilon: 3,
    minPoints: 2,
  );

  List<List<int>> clusterOutput = dbscan.run(dataset);
  print("===== 1 =====");
  print("Clusters output");
  print(clusterOutput);//or dbscan.cluster
  print("Noise");
  print(dbscan.noise);
  print("Cluster label for points");
  print(dbscan.label);

  List<List<int>> clusterOutput2 = dbscan.run(dataset2);
  print("===== 2 =====");
  print("Clusters output");
  print(clusterOutput2);//or dbscan.cluster
  print("Noise");
  print(dbscan.noise);
  print("Cluster label for points");
  print(dbscan.label);
}


example/hierarchical_sample.dart
import 'package:simple_cluster/src/hierarchical.dart';

void main(){
  List<List<double>> colors = [
    [20, 20, 80],
    [22, 22, 90],
    [250, 255, 253],
    [100, 54, 255]
  ];

  List<List<double>> singles = [[7], [10], [20], [28], [35]];

  Hierarchical hierarchical = Hierarchical(
    minCluster: 2, //stop at 2 cluster
    linkage: LINKAGE.SINGLE
  );

  List<List<int>> clusterList = hierarchical.run(colors);

  print("===== 1 =====");
  print("Clusters output");
  print(clusterList);//or hierarchical.cluster
  print("Noise");
  print(hierarchical.noise);
  print("Cluster label for points");
  print(hierarchical.label);

  List<List<int>> clusterList2 = hierarchical.run(singles);

  print("===== 2 =====");
  print("Clusters output");
  print(clusterList2);//or hierarchical.cluster
  print("Noise");
  print(hierarchical.noise);
  print("Cluster label for points");
  print(hierarchical.label);
}